/*
** xrouted message definitions
*/

package Xroute;
//syntax = "proto2";

enum constants {
	XROUTE_PROTO_VERSION = 1;
}

message XID {
	required uint32 type = 1; // should use the XID type values from xia.h
	required bytes  id   = 2; // must be 20 bytes long
}

message Node {
	required XID    ad    = 1;
	required XID    hid   = 2;
	optional XID    sid   = 3;
	optional uint32 flags = 4;
}

message TableEntry {
	required XID    xid       = 1;
	required XID    next_hop  = 2;
	required uint32 interface = 3;
	required uint32 port      = 4;
	optional uint32 flags     = 5;
}

message HelloMsg {
	required Node   node  = 1;
	optional uint32 flags = 2;
}

message LSAMsg {
	required Node   node  = 1;
	optional uint32 flags = 2;
	repeated Node   peers = 3;
}

message HostJoinMsg {
	required string hid       = 1;
	required uint32 interface = 2;
	optional uint32 flags     = 3;
}

// FIXME: make this use the XID type at some point
message HostLeaveMsg {
	required string hid = 1;
}

message TableUpdateMsg {
	required Node       from   = 1;
	required Node       to     = 2;
	repeated TableEntry routes = 3;
}

message GlobalLSAMsg {
	required Node from = 1;
}

message ConfigMsg {
	required string ad             = 1;
	required string controller_dag = 2;
}

enum msg_type {
	CONFIG_MSG              = 1; // router configuration info from xnetjd
	HELLO_MSG               = 2; // local router hello msg. still needed?
	LSA_MSG                 = 3; // router announcement -> controller
	HOST_JOIN_MSG           = 4; // add host route from xnetjd
	HOST_LEAVE_MSG          = 5; // host leaving network
	TABLE_UPDATE_MSG        = 6; // new tables from controller -> router
	GLOBAL_LSA_MSG          = 7; // lsa between controllers in different ADs
	SID_DISCOVERY_MSG       = 8;
	SID_ROUTING_TABLE_MSG   = 9;
	SID_MANAGE_KA_MSG       = 10; // keep alive msg to service controller
	SID_RE_DISCOVERY_MSG    = 11; // send to service controller
	AD_PATH_STATE_PING_MSG  = 12;
	AD_PATH_STATE_PONG_MSG  = 13;
	SID_DECISION_QUERY_MSG  = 14;
	SID_DECISION_ANSWER_MSG = 15;
}

message XrouteMsg {
	required uint32   version = 1;
	required msg_type type    = 2;

	oneof payload {
		HelloMsg       hello        = 3;
		LSAMsg         lsa          = 4;
		TableUpdateMsg table_update = 5;
		GlobalLSAMsg   global_lsa   = 6;
		HostJoinMsg    host_join    = 7;
		HostLeaveMsg   host_leave   = 8;
		ConfigMsg      config       = 9;
	}
}
